New IteratorMincClassCastBugIT and fixes to #1411

milleruntime opened the pull request:
* Create new IT for testing runtime bug that was introduced when moving
the iterators to declare them public API in #1411. The other changes in
this commit are for fixing the bug.
* Drop new class InterruptibleMapIterator that was created in #1411
* Revert SortedMapIterator back to previous implementation where it
was an InterruptibleIterator but keep it in core.iterators pkg
* Move InterruptibleIterator interface to API and add javadoc
* Move ColumnFamilySkippingIterator to API
* Add javadoc to IterationInterruptedException
* Fixes #2341

Manno15 commented on test/src/main/java/org/apache/accumulo/test/functional/IteratorMincClassCastBugIT.java:
```diff
@@ -0,0 +1,516 @@
+      md5digest.reset();
+      InputStream fis = null;
+      int numRead = 0;
+      // try (var reader = new InputStreamReader(accumuloPropsLocation.openStream(), UTF_8)) {
```
```suggestion

```

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iterators/InterruptibleIterator.java:
```diff
@@ -16,14 +16,22 @@
+ * Allows an iterator to be interrupted. Typically, once the interrupt flag is set the iterator will
+ * throw an {@link InterruptedException} if the interrupt is detected. Some iterators have been
+ * optimized to not always check the flag.
+ *
+ * One example of a system interrupt is when a Tablet is being closed. If a Tablet has an active
```
Blank line in javadoc without a `<p>` isn't correct

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iterators/ColumnFamilySkippingIterator.java:
```diff
@@ -16,7 +16,7 @@
  * under the License.
  */
-package org.apache.accumulo.core.iteratorsImpl.system;
+package org.apache.accumulo.core.iterators;
```
Moving the interface back to public API seems like the obvious fix, but are we expecting users to actually use this interface? I haven't followed all the details, but my understanding was that this was a feature intended solely for internal use with some of our own system iterators. But, maybe I'm wrong? Is this something that is useful to user iterators and safe for them to use?

milleruntime replied:
That is what I had originally thought but after investigating this bug it seems like `SortedMapIterator` and this one could be cast to an InterruptibleIterator in the `LocalityGroupIterator`
https://github.com/apache/accumulo/blob/8d33bc4d489880f15950189d20287262ce46e68e/core/src/main/java/org/apache/accumulo/core/iteratorsImpl/system/LocalityGroupIterator.java#L49

I think if a user extends one of them and configures it to run in certain situations it could be created through deepCopy(). But it also seems dangerous either way.

milleruntime replied:
Also, my original intention in #1411 was since SortedMapIterator was used so much in our Tests that it would be useful to users as well. But I could be wrong in that thinking. It could be unsafe to expose it to users so a better option may be to move it in with the other system iterators. I can't remember if I attempted this or how difficult this would be though.

milleruntime replied:
Moved this iterator back to iteratorsImpl package in 1a4bb21607324ef6a9d2a0eeb458c0cdb3c59681

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iterators/SortedMapIterator.java:
```diff
@@ -36,28 +37,40 @@
   private SortedMap<Key,Value> map;
   private Range range;
 
+  private AtomicBoolean interruptFlag;
```
Atomic references are typically `final` whenever possible, to prevent bugs where atomicity is broken because the reference is changed. I see this can't be made final because there's a setter... but that seems like bad API design. Is there something we can do to fix this?

milleruntime replied:
We should be able to make this final with an initial value of `false`. I haven't really touched the iterators since most of them have been around since forever but I think this would be a good fix. FYI I had a hell of a time tracking down where we actually do the interrupt. As a reference, it seems like the only place we actually interrupt iterators is in `ScanDataSource`:
https://github.com/apache/accumulo/blob/8a636a3ba91f5dae1d8b09b095178889a7d79c1d/server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/ScanDataSource.java#L256-L258
When a tablet is closed:
https://github.com/apache/accumulo/blob/b5ca15806659efbbd6f5d158be8648cc2cd13593/server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java#L1288

We have a ton of calls to pass the flag around between iterators, readers and RFile code but that seems to be the only place where the actual interrupt takes place.

milleruntime replied:
I thought about changing the `AtomicBoolean` to be final but now I am wondering if it was written this way with performance in mind. We do a lot of passing around of the flag across the iterator stack but only ever set it in one place.  If we were to make it final, then I think that would be a lot of extra memory accesses to the atomic value vs just passing around a reference.

I thought of doing something like this in the setter, but like I said, I think it would be less efficient:
<pre>
public void setInterruptFlag(AtomicBoolean flag) {
    boolean newFlag = flag.get();
    boolean previousFlag = this.interruptFlag.getAndSet(newFlag);
    if (previousFlag == newFlag && newFlag)
      log.warn("Interrupt called on iterator that was previously interrupted.");
  }
</pre>

ctubbsii replied:
It's not a performance thing. It's an implementation injection mechanism (see where we're notifying the iterators by changing the value in the scan code: `grep -lR 'interruptFlag.set(true)' .`. AtomicBoolean is being used like a BooleanSupplier. This was probably fine at one point, since BooleanSupplier was introduced in Java 8, and before that, AtomicBoolean was an attractive alternative to a built-in supplier. But, having the type be AtomicBoolean implies atomicity, and that can be confusing, since atomicity guarantees are clearly broken by having a setter that can swap out the reference itself in addition to the value. We're not actually relying on the atomicity guarantees in these cases, though, as far as I can tell, so BooleanSupplier would suffice as the type.

Your code above wouldn't work. I had a similar idea initially (`this.interruptFlag.set(flag.get())`). However, it's not sufficient to set the value. We need to set the containing object reference, so we can communicate the change in value when the caller updates the source.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iterators/SortedMapIterator.java:
```diff
@@ -95,6 +111,9 @@ public void next() throws IOException {
   public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive)
       throws IOException {
 
+    if (interruptFlag != null && interruptFlag.get())
```
`interruptFlag` should never be null. We can require non-null anywhere we accept it.

ctubbsii commented on test/src/main/java/org/apache/accumulo/test/functional/IteratorMincClassCastBugIT.java:
```diff
@@ -0,0 +1,516 @@
+    }
+
+    private void insert(ByteSequence cv) {
+      for (int i = 0; i < cv.length(); i++) {
+
+      }
```
What is this loop doing?

ctubbsii commented on test/src/main/java/org/apache/accumulo/test/functional/IteratorMincClassCastBugIT.java:
```diff
@@ -0,0 +1,516 @@
+
+  public static final byte[] nullbyte = new byte[] {0};
+
+  /**
+   * Join some number of strings using a null byte separator into a text object.
+   *
+   * @param s
+   *          strings
+   * @return a text object containing the strings separated by null bytes
+   */
```
This could probably just be a one-line inline comment. We're not publishing javadocs for ITs. This method can probably also be private... as could some others added in this file for testing.

ctubbsii commented on test/src/main/java/org/apache/accumulo/test/functional/IteratorMincClassCastBugIT.java:
```diff
@@ -0,0 +1,516 @@
+        sb.append(cvs);
+      }
+
+      return sb.toString().getBytes();
```
Should be `getBytes(UTF_8)`

ctubbsii commented on test/src/main/java/org/apache/accumulo/test/functional/IteratorMincClassCastBugIT.java:
```diff
@@ -0,0 +1,516 @@
+    Text t = new Text(s[0]);
+    for (int i = 1; i < s.length; i++) {
+      t.append(nullbyte, 0, 1);
+      t.append(s[i].getBytes(), 0, s[i].length());
```
should be `getBytes(UTF_8)`

ctubbsii commented on test/src/main/java/org/apache/accumulo/test/functional/IteratorMincClassCastBugIT.java:
```diff
@@ -0,0 +1,516 @@
+      InputStream fis = null;
+      int numRead = 0;
+      // try (var reader = new InputStreamReader(accumuloPropsLocation.openStream(), UTF_8)) {
+      try {
+        fis = new FileInputStream(filename);
+        numRead = fis.read(buf);
+        while (numRead >= 0) {
+          if (numRead > 0) {
+            md5digest.update(buf, 0, numRead);
+          }
+          numRead = fis.read(buf);
+        }
+      } finally {
+        if (fis != null) {
+          fis.close();
+        }
+      }
```
Our test file is small enough that this can be reduced to two simple lines... one to read the file fully into a String, and another to compute the digest from the String's bytes. This would improve the comprehensibility of the test, since we don't actually need to chunk and buffer the tiny test file.

milleruntime replied:
I rewrote the IT so it doesn't need to ingest a file now.

milleruntime commented:
@Manno15 @ctubbsii thanks for taking a look at this. I guess I should have been more explicit or made this a draft PR because a lot of the code in your feedback was just yanked from the examples repo, specifically the [ChunkCombiner](https://github.com/apache/accumulo-examples/blob/main/src/main/java/org/apache/accumulo/examples/filedata/ChunkCombiner.java) to try and simulate what Datawave is doing with iterators. I am not sure if I should clean it up or just stick it in a test jar so we don't have to look at it. 

Do either of you know how to get MiniAccumuloCluster to load classes from a jar during runtime? I thought it might be a better test to do that but couldn't get it working.

milleruntime commented:
@ctubbsii @Manno15 This PR changed significantly since your initial reviews. I moved the iterators back to the iteratorsImpl package, along with moving SortedMapIterator there. The more I thought about it, the less I liked including InterruptibleIterator in the public API. We created the `YieldingKeyValueIterator` interface with the intention of public use so we don't need InterruptibleIterator in the API. I also rewrote the IT to just use a simple iterator that will still cause the class cast exception to happen without the fix.

Manno15 reviewed (approved):
Changes look good, I don't fully know the impact of the interuptible changes but I didn't run into any issues while testing. Confirmed that these changes to fix the reported issue.

milleruntime commented:
> Changes look good, I don't fully know the impact of the interuptible changes but I didn't run into any issues while testing. Confirmed that these changes to fix the reported issue.

Thanks for testing these changes. Just curious, how did you reproduce the issue?

Manno15 commented:
> Thanks for testing these changes. Just curious, how did you reproduce the issue?

Using the IT you created. So maybe that isn't a definitive test but it did fail against main while working with the rest of these changes.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iterators/IterationInterruptedException.java:
```diff
@@ -18,6 +18,9 @@
+/**
+ * Exception thrown if an interrupt flag is detected.
+ */
 public class IterationInterruptedException extends RuntimeException {
```
Should this RTE exist in public API, or should it be in iteratorsImpl as well?

milleruntime replied:
Good catch. I moved it to iteratorsImpl in 892a867. Thankfully, nothing outside of the tserver code and Rfile use it.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iteratorsImpl/system/SortedMapIterator.java:
```diff
@@ -121,4 +143,9 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
     throw new UnsupportedOperationException();
   }
+
+  @Override
+  public void setInterruptFlag(AtomicBoolean flag) {
+    this.interruptFlag = flag;
+  }
```
Because changing the reference breaks atomicity guarantees, this method looks dubious. It looks like the way this is intended to be used is like a BooleanSupplier (or even more strictly, a one-way latch that can move from false to true, but never back to false). I'm not sure the right solution, but if the interruptible stuff is only on the server-side, it could be cleaned up to avoid using AtomicBoolean and use something more appropriate instead.

milleruntime replied:
I agree. I am hesitant to go changing internals with the iterators though since they are so heavily reliant on runtime configuration (one of the reasons it took so long to find this bug in the first place) and I don't think we have adequate test coverage for their behavior.

ctubbsii replied:
My suggestion would be to use BooleanSupplier in the InterruptibleIterator interface, and call `setInterruptFlag(interruptFlag::get)` (where `interruptFlag` in the caller is of type `AtomicBoolean`). That would remove any confusion over implied atomicity guarantees inside the InterruptibleIterator interface, but wouldn't change any behavior. But, obviously, that would be a follow-on issue. It need not be done as part of this PR.

ctubbsii reviewed (approved):
I like the consolidation of InterruptibleMapIterator into SortedMapIterator, and I like keeping these out of the public API. I'm not sure I understand how it fixed the reported issue, but these changes look okay to me.
